torchvision/models/unet.py

```diff
@@ -0,0 +1,154 @@
+import torch
+import torch.nn as nn
+
+
+__all__ = ['UNet', 'unet8', 'unet13', 'unet18', 'unet23', 'unet28', 'unet33']
+
+
+def double_conv(in_channels, out_channels):
+    return nn.Sequential(
+        nn.Conv2d(in_channels, out_channels, 3),
+        nn.ReLU(inplace=True),
+        nn.Conv2d(out_channels, out_channels, 3),
+        nn.ReLU(inplace=True)
+    )
+
+
+def center_crop(img, output_size):
+    _, _, h, w = img.size()
+    _, _, th, tw = output_size
+    i = (h - th) // 2
+    j = (w - tw) // 2
+    return img[:, :, i:i + th, j:j + tw]
+
+
+class Contract(nn.Module):
+
+    def __init__(self, in_channels, out_channels, dropout=False, p=0.5):
+        super(Contract, self).__init__()
+        assert in_channels < out_channels
+
+        self.pool = nn.MaxPool2d(2)
+        self.conv = double_conv(in_channels, out_channels)
+        self.drop = None
+
+        if dropout:
+            self.drop = nn.Dropout2d(p=p)
+
+    def forward(self, x):
+        x = self.pool(x)
+        x = self.conv(x)
+
+        if self.drop is not None:
+            x = self.drop(x)
+
+        return x
+
+
+class Expand(nn.Module):
+
+    def __init__(self, in_channels, out_channels):
+        super(Expand, self).__init__()
+        assert in_channels > out_channels
+
+        self.upconv = nn.ConvTranspose2d(in_channels, out_channels, 2, 2)
+        self.relu = nn.ReLU(inplace=True)
+        self.conv = double_conv(in_channels, out_channels)
+
+    def forward(self, x, out):
+        x = self.upconv(x)
+        x = self.relu(x)
+
+        out = center_crop(out, x.size())
+        x = torch.cat([out, x], 1)
+
+        x = self.conv(x)
+
+        return x
+
+
+class UNet(nn.Module):
+    """`U-Net <https://arxiv.org/pdf/1505.04597.pdf>`_ architecture.
+
+    Args:
+        in_channels (int, optional): number of channels in input image
+        out_channels (int, optional): number of channels in output segmentation
+        start_channels (int, optional): power of 2 channels to start with
+        depth (int, optional): number of contractions/expansions
+        p (float, optional): dropout probability
+    """
+
+    def __init__(self, in_channels=1, out_channels=2, start_channels=6,
```

Comment:
Does `out_channels` here correspond to the number of heatmaps, each representing a heatmap for an individual class?If yes then you might consider renaming it to `num_classes` following the naming convention of pytorch models.

Reply:
Yes, that's what it represents. I went with `out_channels` because there didn't seem to be a consensus among existing models:
```console
$ grep '__init__.*num_classes' *
alexnet.py:    def __init__(self, num_classes=1000):
deeplabv3.py:    def __init__(self, in_channels, num_classes):
googlenet.py:    def __init__(self, num_classes=1000, aux_logits=True, transform_input=False, init_weights=True):
googlenet.py:    def __init__(self, in_channels, num_classes):
inception.py:    def __init__(self, num_classes=1000, aux_logits=True, transform_input=False):
inception.py:    def __init__(self, in_channels, num_classes):
mobilenet.py:    def __init__(self, num_classes=1000, width_mult=1.0):
resnet.py:    def __init__(self, block, layers, num_classes=1000, zero_init_residual=False,
shufflenetv2.py:    def __init__(self, stages_repeats, stages_out_channels, num_classes=1000):
squeezenet.py:    def __init__(self, version='1_0', num_classes=1000):
vgg.py:    def __init__(self, features, num_classes=1000, init_weights=True):
$ grep '__init__.*out_channels' *
deeplabv3.py:    def __init__(self, in_channels, out_channels, dilation):
deeplabv3.py:    def __init__(self, in_channels, out_channels):
googlenet.py:    def __init__(self, in_channels, out_channels, **kwargs):
inception.py:    def __init__(self, in_channels, out_channels, **kwargs):
shufflenetv2.py:    def __init__(self, stages_repeats, stages_out_channels, num_classes=1000):
unet.py:    def __init__(self, in_channels, out_channels, dropout=False, p=0.5):
unet.py:    def __init__(self, in_channels, out_channels):
unet.py:    def __init__(self, in_channels=1, out_channels=2, start_channels=6,
```
I'll switch to `num_classes`.

Reply:
Okay, switched it to `num_classes`. The helper functions still use `out_channels` since that number is independent of the final number of classes at the end.